src/envoy/http/authn/authenticator_base.h

```diff
@@ -0,0 +1,82 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "authentication/v1alpha1/policy.pb.h"
+#include "common/common/logger.h"
+#include "server/config/network/http_connection_manager.h"
+#include "src/envoy/http/authn/context.pb.h"
+#include "src/envoy/http/authn/filter_context.h"
+
+namespace Envoy {
+namespace Http {
+namespace IstioAuthN {
+
+// AuthenticatorBase is the base class for authenticator. It provides functions
+// to perform individual authentication methods, which can be used to construct
+// compound authentication flow.
+class AuthenticatorBase : public Logger::Loggable<Logger::Id::filter> {
+ public:
+  // Callback type for individual authentication method.
+  typedef std::function<void(const Payload*, bool)> MethodDoneCallback;
+
+  // Callback type for the whole authenticator.
+  typedef std::function<void(bool)> DoneCallback;
+
+  AuthenticatorBase(FilterContext* filter_context,
+                    const DoneCallback& callback);
+  virtual ~AuthenticatorBase();
+
+  // Perform authentication.
+  virtual void run() PURE;
+
+  // Calls done_callback_ with success value.
+  void done(bool success) const;
+
+  // Validates x509 given the params (more or less, just check if x509 exists,
+  // actual validation is not neccessary as it already done when the connection
+  // establish), and extract authenticate attributes (just user/identity for
+  // now). Calling callback with the extracted payload and corresponding status.
+  virtual void validateX509(
+      const istio::authentication::v1alpha1::MutualTls& params,
+      const MethodDoneCallback& done_callback) const;
```

Comment:
We should go with Envoy coding style.   Instead of using std::function,  it prefers to define abstract callback base class

class Callback {
     virutal onDone() PURE;
}

Reply:
Is this some strict style? I'd like to do in the next PRs.

On the other hand, I personally think using std::function for callback is cleaner, and easier to write test.

Reply:
No, I don't plan to upstream this. It's very Istio specific.

Btw, for the defense, I think callback using std::function is not a rare thing in envoy codebase

https://github.com/envoyproxy/envoy/search?utf8=%E2%9C%93&q=std%3A%3Afunction+callback&type=